Set-Cookie headers should not be combined

madgnome opened the pull request:
Motivation:

According to the HTTP spec set-cookie headers should not be combined
because they are not using the list syntax.

Modifications:

Added a blacklist to CombinedHttpHeaders, headers in the blacklist are
not combined. The default blacklist contains Set-Cookie.

Result:

Set-Cookie headers won't be combined anymore
Fixes #6473

normanmaurer commented:
@netty-bot test this please

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -80,14 +97,16 @@ public CharSequence escape(CharSequence value) {
         public CombinedHttpHeadersImpl(HashingStrategy<CharSequence> nameHashingStrategy,
                 ValueConverter<CharSequence> valueConverter,
-                io.netty.handler.codec.DefaultHeaders.NameValidator<CharSequence> nameValidator) {
+                io.netty.handler.codec.DefaultHeaders.NameValidator<CharSequence> nameValidator,
```
@madgnome can we please keep the original constructor as well ?

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -36,8 +39,21 @@
+        NON_COMBINABLE_HEADERS = new HashSet<CharSequence>();
+
+        // According to the spec Set-cookie header does not use the list syntax and cannot be combined
+        NON_COMBINABLE_HEADERS.add(SET_COOKIE);
```
please store this as non-mutable set

ejona86 replied:
Could we just instead special-case this one key (with a plain condition)?

normanmaurer commented:
@slandelle PTAL as well

slandelle commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -213,9 +232,13 @@ public CombinedHttpHeadersImpl setObject(CharSequence name, Iterable<?> values)
             return this;
         }
 
+        private boolean canBeCombined(CharSequence name) {
```
You always use the opposite `!canBeCombined` so IMHO it would be better to have a `cantBeCombined` instead (I would expect IntelliJ to complain as well).

madgnome replied:
Oh yeah good point, at some point I was using `canBeCombined` without the negation and wanted to avoid the double negative but it's gone now, so will change it.

slandelle commented:
LGTM except for a small nit.

slandelle commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -213,9 +232,13 @@ public CombinedHttpHeadersImpl setObject(CharSequence name, Iterable<?> values)
         }
 
+        private boolean canBeCombined(CharSequence name) {
+            return !nonCombinableHeaders.contains(name);
```
If I'm not missing something, this should be case insensitive. You should probably be using the hasher instead.

normanmaurer replied:
Thats a very good point...  @madgnome please also add a unit test for this.

ejona86 commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -36,8 +39,21 @@
+        super(new CombinedHttpHeadersImpl(CASE_INSENSITIVE_HASHER, valueConverter(validate), nameValidator(validate), NON_COMBINABLE_HEADERS));
+    }
+
+    public CombinedHttpHeaders(boolean validate, Set<CharSequence> nonCombinableHeaders) {
```
Adding this constructor seems like an overkill, to let users provide their own new headers-that-are-broken since [HTTP itself only specifies Set-Cookie as an exception](https://tools.ietf.org/html/rfc7230#section-3.2.2).

madgnome commented:
I've actioned the feedback:

- Removed Set<> and just special cased `Set-Cookie`
- renamed to `cannotCombined` since we were alway using the negative form
- Using `CASE_INSENSITIVE_HASHER` to do the comparison

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java:
```diff
@@ -213,9 +215,13 @@ public CombinedHttpHeadersImpl setObject(CharSequence name, Iterable<?> values)
         }
 
+        private static boolean cannotBeCombined(CharSequence name) {
+            return CASE_SENSITIVE_HASHER.equals(SET_COOKIE, name);
```
I think you could just use `SET_COOKIE.contentEqualsIgnoreCase(...)`.

madgnome replied:
Fixed!

normanmaurer commented:
@netty-bot test this please
